### PR TITLE
compositing: Implement display ports and avoid creating display lists for items outside it.

### PR DESCRIPTION
**Pull request by pcwalton**
This improves Servo's performance on large pages.

Please double-check the logic when it comes to nested layers—I'm sure I've messed up some of the geometry calculations :)

r? @glennw

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6053)

### CODE REVIEW

File: components/compositing/compositor.rs [hunk 1]
```diff
@@ -937,9 +939,59 @@ impl<Window: WindowMethods> IOCompositor<Window> {
...
+
+            for kid in layer.children.borrow().iter() {
+                process_layer(&*kid, window_size, new_display_ports)
+            }
+        }
+
+        let dppx = (self.page_zoom * self.device_pixels_per_screen_px()).get();
+        let window_size = self.window_size.as_f32().to_untyped();
+        let window_size = Size2D(window_size.width / dppx, window_size.height / dppx);
+        let window_size = TypedSize2D::from_untyped(&window_size);
```
Comment (mbrubeck) [inline components/compositing/compositor.rs]: You could write this without any untyped units:

```rust
        let dppx = self.page_zoom * self.device_pixels_per_screen_px();
        let window_size = self.window_size.as_f32() / dppx * ScaleFactor::new(1.0);
```

The final ScaleFactor is from ViewportPx to LayerPixel.

**Comment by glennw** [comment]
Review status: 19 of 24 files reviewed, 2 unresolved discussions, some commit checks failed.
Reviewed files:
- components/compositing/pipeline.rs @ r1
- components/gfx/paint_task.rs @ r1
- components/layout/context.rs @ r1
- components/layout/flow.rs @ r1
- components/layout/list_item.rs @ r1
- components/layout/multicol.rs @ r1
- components/layout/table.rs @ r1
- components/layout/table_caption.rs @ r1
- components/layout/table_cell.rs @ r1
- components/layout/table_row.rs @ r1
- components/layout/table_rowgroup.rs @ r1
- components/layout/table_wrapper.rs @ r1
- components/layout/traversal.rs @ r1
- components/layout_traits/Cargo.toml @ r1
- components/layout_traits/lib.rs @ r1
- components/msg/compositor_msg.rs @ r1
- components/script/layout_interface.rs @ r1
- components/servo/Cargo.lock @ r1
- tests/html/lipsum-large.html @ r1

---

<sup>**[components/servo/Cargo.lock, line 676 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6053#-JpKkPyevMXl_joKdd_P)** ([raw file](https://github.com/servo/servo/blob/52fc100bc6ba0a50d0e02384878a3ed3610b01e2/components/servo/Cargo.lock#L676)):</sup>
Need to run update-cargo to update gonk/ and cef/ 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6053)

**Comment by glennw** [comment]
@pcwalton While looking at this I'm wondering how we can reftest it. What about a special compositor mode that can be set in reftests which fixes the display ports to subregion(s) of the visible screen, and we can write reftests to verify that items are either culled or visible based on their location? That wouldn't test everything but would test the error prone parts of the culling code.

It doesn't need to be done for this PR, but it could be a good follow up. Do you think that could work or have you got other ideas on how to reftest this?

**Comment by pcwalton** [comment]
That sounds like a good idea. We could have a debug clipping flag that uses this infrastructure to clip to a specific region of the screen.

**Comment by pcwalton** [comment]
Fixed tidy.

**Comment by glennw** [comment]
Review status: all files reviewed, 3 unresolved discussions, all commit checks successful.
Reviewed files:
- components/compositing/compositor.rs @ r2
- components/gfx/paint_task.rs @ r2
- components/layout/block.rs @ r2
- components/layout/display_list_builder.rs @ r2
- components/layout/inline.rs @ r1
- components/layout/layout_task.rs @ r1
- components/servo/Cargo.lock @ r2

---

<sup>**[components/layout/block.rs, line 1785 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6053#-JpZ8bgrw9zH7eUx5oDI)** ([raw file](https://github.com/servo/servo/blob/009dd182e06bc8306208cd47106693928e3b35bb/components/layout/block.rs#L1785)):</sup>
Point2D::zero()

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6053)

**Comment by glennw** [comment]
@pcwalton Left a few comments on reviewable. The geometry calculations seem right to me. I've verified that all the ref/wpt tests pass and also checked browsing the regular sites I test on with this branch, and they seem fine.

One other question - does this work when you click on a fragment link (i.e. instantly change the viewport to a different area of the document?).

Once those questions / comments are addressed, looks good to me.

**Comment by pcwalton** [comment]
@bors-servo: r=glennw

I've tested scrolling to a fragment point and added `tests/html/test_scroll_to_point.html` to make sure this works.

**Pull request merged**